proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+2. [Retreiving the info table pointer: primtype `InfoTablePtrOf#` and compile-time primop `reifyInfoTablePtr#`](#retrieving-the-info-table-pointer-primtype-infotableptrof-and-compile-time-primop-reifyinfotableptr)
+3. [Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family](#obtain-data-constructor-name-as-a-symbol-liftedctortosymbol-built-in-type-family)
+
+### Allocate hollow data constructor in region: RTS primop `compactAddHollow#`
+
+```haskell
+compactAddHollow#
+  :: Compact#         -- ^ the compact region where the constructor heap object will be added
+  -> Addr#            -- ^ the info table pointer that will be set as header for the new constructor heap object
+  -> State# RealWorld -- ^ state token
+  -> (# State# RealWorld
+      , a             -- ^ the allocated constructor heap object
+      , Word#         -- ^ size of the header of the allocated heap object, in bytes
+      , Word#         -- ^ size of the words, in bytes, on the platform
+     #)
+```
+
+This primitive must be implemented in the RTS (in `rts/Compact.cmm`). The behavior of its Cmm code should correspond to these macro steps:
+
+```c
+stg_compactAddHollowzh(
+  P_ compact, // the compact region where the constructor heap object will be added
+  W_ info     // the info table pointer that will be set as header for the new constructor heap object
+) {
+  [...]
+  // find address where to store the pointer to the new constructor object
+  pp = compact + SIZEOF_StgHeader + OFFSET_StgCompactNFData_result;
+  // reuse existing macro to allocate inside the compact region
+  ALLOCATE(compact, size, p, new_ctor_obj, tag);
+  // write the supplied info table pointer
+  SET_HDR(new_ctor_obj, info, CCS_SYSTEM);
+  // store the pointer to the new constructor in the return location
+  P_[pp] = new_ctor_obj;
+  [...]
+
+  return (P_[pp], SIZEOF_StgHeader, SIZEOF_W);
+}
+```
+
+`compactAddHollow#` allocates space for a new constructor heap object in the compact region, but doesn't set any of its fields (so it is _hollow_). It only sets its header with the supplied info table pointer. The user is responsible for setting the fields of the constructor heap object before reading it. The primitive returns the newly allocated constructor heap object, but also the size of the header of the object, and the size of the words on the platform so that the user can later set the fields of the returned object.
+
+### Retrieving the info table pointer: primtype `InfoTablePtrOf#` and compile-time primop `reifyInfoTablePtr#`
+
+`compactAddHollow#` is in charge of allocating the constructor heap object in the compact region, but it doesn't have access to the info table pointer for the constructor. All information about info tables has long been inlined/erased at runtime. So we need a primitive `reifyInfoTablePtr#` to reify the info table pointer of a given constructor at compile-time into a static value that the RTS can then use.
```

Comment:
Not “is in charge” tell us what it does/means. This section is a specification. Explanations of how these primitive interact go in the _Example_ and _Effects and interactions_ section.

Reply:
I moved it. I hope that the readers will still understand why I propose this specific set of changes. It's a bit weird to justify the changes  much after I first specify them.